src/Tools/dotnet-monitor/Exceptions/ConfiguredExceptionsStore.cs

```diff
@@ -0,0 +1,149 @@
+            // access to this relationship without having to look up IExceptionInstances.
+            private readonly Dictionary<ulong, List<ulong>> _innerExceptionMap = new();
+            // Track the list of outer exceptions for a given exception; allows for quicker
+            // access to this relationship without having to look up IExceptionInstances.
+            private readonly Dictionary<ulong, List<ulong>> _outerExceptionMap = new();
+            private readonly ConfiguredExceptionsStore _store;
+            // Front of the list has the latest top-level exception ID and the
+            // back of the list has the oldest top-level exception ID.
+            private readonly LinkedList<ulong> _topLevelExceptions = new();
+            private readonly int _topLevelLimit;
+
+            public Callback(ConfiguredExceptionsStore store, int topLevelLimit, ExceptionsStoreCallback callback)
+            {
+                _callback = callback;
+                _store = store;
+                _topLevelLimit = topLevelLimit;
+            }
+
+            public override void AfterAdd(IExceptionInstance instance)
+            {
+                _callback?.AfterAdd(instance);
+            }
+
+            public override void BeforeAdd(IExceptionInstance instance)
+            {
+                _outerExceptionMap.Add(instance.Id, new List<ulong>());
+                _innerExceptionMap.Add(instance.Id, instance.InnerExceptionIds.ToList());
+
+                // Make sure all of the inner exceptions are no longer top-level exceptions.
+                foreach (ulong innerExceptionId in instance.InnerExceptionIds)
+                {
+                    if (_outerExceptionMap.TryGetValue(innerExceptionId, out List<ulong> outerExceptions))
+                    {
+                        outerExceptions.Add(instance.Id);
+                    }
+
+                    _topLevelExceptions.Remove(innerExceptionId);
+                }
+
+                _topLevelExceptions.AddFirst(instance.Id);
+
+                // If over the limit, remove the oldest exception as well as its inner
+                // exceptions that are not shared with any current top-level exception.
+                if (_topLevelExceptions.Count > _topLevelLimit)
+                {
+                    LinkedListNode<ulong> lastNode = _topLevelExceptions.Last;
+                    _topLevelExceptions.RemoveLast();
+
+                    RemoveIfNoOuterExceptions(lastNode.Value);
+                }
+
+                _callback?.BeforeAdd(instance);
+            }
+
+            private void RemoveIfNoOuterExceptions(ulong exceptionId)
+            {
+                // Remove the exception if it no longer belongs to any other exception in the store
+                // as an inner exception.
+                if (_outerExceptionMap.TryGetValue(exceptionId, out List<ulong> outerExceptionIds))
```

Comment:
I know we had previously discussed this briefly but I don't think I saw it in here (sorry if I missed it) - do we want any mechanism for limiting inner-exception depth (i.e. we only have 20 top-level exceptions, but they go 1000 inner exceptions deep). Not saying that limiting needs to be added, just wanted to make sure it was brought up.

Reply:
That's something I can add as a separate change (don't want to complicate this one further).

I think the easiest limit to explain to customers (but not necessarily the easiest to implement) is to say that every exception can only have a certain inner exception depth and everything outside of that depth would be eligible for removal.